adr/2020_05_13_handling_non_crud_operations_in_api.md

```diff
@@ -0,0 +1,132 @@
+# Handling Non-CRUD operations in API
+
+* Status: proposed
+* Date: 2020-05-13
+
+## Context and Problem Statement
+
+Handling Non-CRUD operation over REST API is not trivial. Once operations are beyond Create, Read, Update, Delete, 
+which can be mapped directly to `POST`/`PUT`, `GET`, `PUT`/`PATCH`, `DELETE` HTTP methods there is no clear recommendation 
+how to map others actions. These actions include, but are not limited to, changes of states (described in the form of 
+workflow in a state machine) or command execution.
+
+## Decision Drivers
+
+* Solution should not be limited by its convention. We may need to support two different actions of the same name, 
+but with different business logic. E.g., One object may have two transitions with the same name but described by 
+two different state machines.
+* Solution should allow for an easy understanding of what is expected behavior of its execution.
+* Solution should provide easy to implement a way of exposing the next possible actions. (according to HATEOAS paradigm).
+* Solution should provide a clean way of adding additional fields during the execution of requested operations.
+
+## Considered Options
+
+All options are presented on the example of making a ship's operations over the shipment, that is ready to be shipped.
+ 
+Expectations: be able to call `ship` transition on the `Shipment` with the XYZ tracking code. 
+
+### Enrich resource data with possible operations
+
+It is possible to add additional fields or reuse current ones to accept operations and leverage HTTP Verbs. In 
+this solution we can use regular `Update` request to execute additional logic:
+
+PUT `/api/orders/1/shipment/1`
+
+```json
+{
+  "status": "ship",
+  "tracking_code": "XYZ"
+}
+```
+
+This solution works for all standards supported by ApiPlatform(JSON-LD, Json HAL, GraphQL)
+
+* Good, because it does not introduce any new `verbs` (which may be considered as a bad practice for REST API)
```

Comment:
Btw. I'm not sure whether we should use "verbs" or "HTTP methods"/"HTTP request methods".